src/js/fable-core/Array2.fs

```diff
@@ -1,16 +1,732 @@
+    result
+
+let private concatImpl (cons: ArrayCons) (arrays: 'T[][]): 'T[] =
+    if arrays.Length > 0 then
+        if isTypedArray arrays.[0] then
+            typedArrayConcatImpl cons arrays
+        else
+            dynamicArrayConcatImpl arrays
+    else
+        cons.Create 0
+
+let concat (arrays: 'T[] seq) (cons: ArrayCons): 'T[] =
+    arrays
+    |> Seq.toArray
+    |> concatImpl cons
+
+let collect (mapping: 'T -> 'U[]) (array: 'T[]) (cons: ArrayCons): 'U[] =
+    map mapping array cons
+    |> concatImpl cons
+
+let countBy (projection: 'T->'Key) (array: 'T[]) (cons: ArrayCons) =
+    let dict = System.Collections.Generic.Dictionary<'Key, int>()
+
+    for value in array do
+        let key = projection value
+        let mutable prev = Unchecked.defaultof<_>
+        if dict.TryGetValue(key, &prev) then dict.[key] <- prev + 1 else dict.[key] <- 1
+
+    let res = cons.Create dict.Count
+    let mutable i = 0
+    for group in dict do
+        res.[i] <- group.Key, group.Value
+        i <- i + 1
+    res
+
+let distinctBy projection (array:'T[]) (cons: ArrayCons) =
+    let temp = cons.Create array.Length
+    let mutable i = 0
+
+    let hashSet = System.Collections.Generic.HashSet<'T>()
+    for v in array do
+        if hashSet.Add(projection v) then
+            temp.[i] <- v
+            i <- i + 1
+
+    getSubArray temp 0 i
+
+let distinct (array: 'T[]) (cons: ArrayCons) = distinctBy id array cons
+
+let where predicate (array: _[]) = filter predicate array
+
+let except (itemsToExclude: seq<_>) (array:_[]) =
+    if array.Length = 0 then
+        array
+    else
+        let cached = System.Collections.Generic.HashSet(itemsToExclude)
+        array |> filter cached.Add
+
+let groupBy (projection: 'T->'Key) (array: 'T[]) (cons: ArrayCons) =
+    let dict = System.Collections.Generic.Dictionary<'Key, 'T[]>()
+
+    // Build the groupings
+    for i = 0 to (array.Length - 1) do
+        let v = array.[i]
+        let key = projection v
+        let mutable prev = Unchecked.defaultof<_>
+        if dict.TryGetValue(key, &prev) then
+            push prev v |> ignore
+        else
+            let prev = cons.From [|v|]
+            dict.[key] <- prev
+
+    // Return the array-of-arrays.
+    let result = cons.Create dict.Count
+    let mutable i = 0
+    for group in dict do
+        result.[i] <- group.Key, group.Value
+        i <- i + 1
+
+    result
+
+let ofArray (arr: 'T[]) = arr
+
+let empty<'T> : 'T[] = [||] // TODO: Shoud it be a ctor function with cons.Create 0?
+
+let inline singleton value (cons: ArrayCons) = cons.From [|value|]
+
+// TODO: CompiledName: Initialize
+let init count initializer (cons: ArrayCons) =
+    if count < 0 then invalidArg "count" LanguagePrimitives.ErrorStrings.InputMustBeNonNegativeString
+    let result = cons.Create count
+    for i = 0 to count - 1 do
+        result.[i] <- initializer i
+    result
+
+[<CompiledName("Pairwise")>]
+let pairwise (array: 'T[]) (cons: ArrayCons) =
```

Comment:
Although it uses the compiled names (e.g. `iterate` instead of `iter`) Fable lowers the first letter of core methods. I did it like that because the core library was originally written in JS and that's the language standards. So in this case we shouldn't need the attribute.

Reply:
Got it!